Support IAccessible2 labelled-by relation

michaelweghorn opened the pull request:
### Link to issue number:

Fixes #17436

### Summary of the issue:

IAccessible2 has an IA2_RELATION_LABELLED_BY relation type, see https://accessibility.linuxfoundation.org/a11yspecs/ia2/docs/html/group__grp_relations.html#ga7bbace7ffb57476b75d621af2e27d1ff . However, that one was not taken into account by NVDA's "labeledBy" property for objects. This could could be seen for example with LibreOffice that implements handling of that IAccessible2 relation.

Additionally, a `None` return value from
IAccessibleHandler.accNavigate wasn't handled
in `IAccessible._get_labeledBy`, triggering an error, e.g.:

    >>> focus.labeledBy
    Traceback (most recent call last):
      File "<console>", line 1, in <module>
      File "C:\tools\cygwin\home\user\development\git\nvda\source\baseObject.py", line 59, in __get__
        return instance._getPropertyViaCache(self.fget)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "C:\tools\cygwin\home\user\development\git\nvda\source\baseObject.py", line 167, in _getPropertyViaCache
        val = getterMethod(self)
              ^^^^^^^^^^^^^^^^^^
      File "C:\tools\cygwin\home\user\development\git\nvda\source\NVDAObjects\IAccessible\__init__.py", line 1167, in _get_labeledBy
        (pacc, accChild) = IAccessibleHandler.accNavigate(
        ^^^^^^^^^^^^^^^^
    TypeError: cannot unpack non-iterable NoneType object

### Description of user facing changes

None, unless they're using the Python console or addons making use of the property.

### Description of development approach

* Extend `IAccessibleHandler.RelationType` with the `LABELLED_BY` relation as defined in the IAccessible2 spec.
* Use this relation in `IAccessible._get_labeledBy` before falling back to trying a custom Mozilla/Gecko specific NAVRELATION.
* Add handling for the case where `IAccessibleHandler.accNavigate` for the custom Mozilla/Geck approach returns `None` to fix the error triggered otherwise when no relation is set.

### Testing strategy:

1. start NVDA
2. Start LibreOffice Writer
3. open the options dialog: "Tools" -> "Options", go to the "User Data" page
4. move focus to the "Company" text edit
5. open NVDA's Python console (Ctrl+Insert+Z)
6. print the object that the currently focused edit is labelled by, and it's accessible name and role:

    >>> focus.labeledBy
    <NVDAObjects.IAccessible.IAccessible object at 0x0BC55A70>
    >>> focus.labeledBy.name
    'Company:'
    >>> focus.labeledBy.role
    <Role.STATICTEXT: 7>

### Known issues with pull request:

None

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
@coderabbitai summary

michaelweghorn committed:
Support IAccessible2 labelled-by relation

 ### Link to issue number:

Fixes #17436

 ### Summary of the issue:

IAccessible2 has an IA2_RELATION_LABELLED_BY relation type, see
https://accessibility.linuxfoundation.org/a11yspecs/ia2/docs/html/group__grp_relations.html#ga7bbace7ffb57476b75d621af2e27d1ff .
However, that one was not taken into account by NVDA's "labeledBy"
property for objects. This could could be seen for example with LibreOffice that
implements handling of that IAccessible2 relation.

Additionally, a `None` return value from
IAccessibleHandler.accNavigate wasn't handled
in `IAccessible._get_labeledBy`, triggering an error, e.g.:

    >>> focus.labeledBy
    Traceback (most recent call last):
      File "<console>", line 1, in <module>
      File "C:\tools\cygwin\home\user\development\git\nvda\source\baseObject.py", line 59, in __get__
        return instance._getPropertyViaCache(self.fget)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "C:\tools\cygwin\home\user\development\git\nvda\source\baseObject.py", line 167, in _getPropertyViaCache
        val = getterMethod(self)
              ^^^^^^^^^^^^^^^^^^
      File "C:\tools\cygwin\home\user\development\git\nvda\source\NVDAObjects\IAccessible\__init__.py", line 1167, in _get_labeledBy
        (pacc, accChild) = IAccessibleHandler.accNavigate(
        ^^^^^^^^^^^^^^^^
    TypeError: cannot unpack non-iterable NoneType object

 ### Description of user facing changes

In NVDA's Python console, retrieving the "labeledBy" property now works for objects in applications implementing the "labelled-by" IAccessible2 relation and no longer triggers an error.

 ### Description of development approach

* Extend `IAccessibleHandler.RelationType` with the `LABELLED_BY` relation
  as defined in the IAccessible2 spec.
* Use this relation in `IAccessible._get_labeledBy` before
  falling back to trying a custom Mozilla/Gecko specific NAVRELATION.
* Add handling for the case where `IAccessibleHandler.accNavigate`
  for the custom Mozilla/Geck approach returns `None` to fix
  the error triggered otherwise when no relation is set.

 ### Testing strategy:

1. start NVDA
2. Start LibreOffice Writer
3. open the options dialog: "Tools" -> "Options", go to the "User Data" page
4. move focus to the "Company" text edit
5. open NVDA's Python console (Ctrl+Insert+Z)
6. print the object that the currently focused edit is labelled by,
   and it's accessible name and role:

    >>> focus.labeledBy
    <NVDAObjects.IAccessible.IAccessible object at 0x0BC55A70>
    >>> focus.labeledBy.name
    'Company:'
    >>> focus.labeledBy.role
    <Role.STATICTEXT: 7>

 ### Known issues with pull request:

None

 ### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
@coderabbitai summary

LeonarddeR reviewed (approved):
Awesome!

seanbudd commented on source/NVDAObjects/IAccessible/__init__.py:
```diff
@@ -1163,12 +1163,19 @@ def isPointInObject(self, x, y):
 		return True
 
 	def _get_labeledBy(self):
```
please add a return type

```suggestion
	def _get_labeledBy(self) -> IAccessible | None:
```

seanbudd commented on user_docs/en/changes.md:
```diff
@@ -70,6 +70,7 @@ Specifically, MathML inside of span and other elements that have the attribute `
 * Opening the NVDA Python Console will no longer fail in case an error occurs while retrieving snapshot variables. (#17391, @CyrilleB79)
 * In Notepad and other UIA documents on Windows 11, if the last line is empty, the `braille next line command` will move the cursor to the last line.
 In any document, if the cursor is on the last line, it will be moved to the end when using this command. (#17251, @nvdaes)
+* In NVDA's Python console, retrieving the "labeledBy" property now works for objects in applications implementing the "labelled-by" IAccessible2 relation and no longer triggers an error. (#17436, @michaelweghorn)
```
This isn't a user facing change, if there's user impact, can you rephrase this to describe it. If there's no direct user impact, please move this to "changes for developers"

michaelweghorn replied:
Done

michaelweghorn committed:
Add type annotation

michaelweghorn committed:
Move changelog entry to "Changes for Developers" section

michaelweghorn committed:
Avoid type clash due to typing.List import

Use qualified name.

Otherwise the newly introduced

    from __future__ import annotations

in

    commit feea4aa0aa0a6b12c33ee8dd70f22d76a2cc6d83
    Author: Michael Weghorn <m.weghorn@posteo.de>
    Date:   Wed Nov 27 09:12:38 2024 +0000

        Add type annotation

causes

    source/NVDAObjects/IAccessible/__init__.py:2391:7: F811 Redefinition of unused `List` from line 13
         |
    2391 | class List(IAccessible):
         |       ^^^^ F811
    2392 |     def _get_role(self):
    2393 |         return controlTypes.Role.LIST
         |
         = help: Remove definition: `List`

    Found 1 error.

seanbudd commented on source/NVDAObjects/IAccessible/__init__.py:
```diff
@@ -1947,7 +1947,7 @@ def _get_detailsRelations(self) -> Tuple["IAccessible"]:
 		return tuple(detailsRelsGen)
 
-	def _get_controllerFor(self) -> List[NVDAObject]:
+	def _get_controllerFor(self) -> typing.List[NVDAObject]:
```
```suggestion
	def _get_controllerFor(self) -> list[NVDAObject]:
```

seanbudd commented on source/NVDAObjects/IAccessible/__init__.py:
```diff
@@ -3,13 +3,13 @@
 # This file is covered by the GNU General Public License.
 # See the file COPYING for more details.
 
+from __future__ import annotations
```
could you comment what this is needed for?

michaelweghorn replied:
I've added a comment above the import. Is it OK like that?

seanbudd commented on user_docs/en/changes.md:
```diff
@@ -100,6 +99,7 @@ Add-ons will need to be re-tested and have their manifest updated.
 * Removed the requirement to indent function parameter lists by two tabs from NVDA's Coding Standards, to be compatible with modern automatic linting. (#17126, @XLTechie)
 * Added the [VS Code workspace configuration for NVDA](https://nvaccess.org/nvaccess/vscode-nvda) as a git submodule. (#17003)
 * In the `brailleTables` module, a `getDefaultTableForCurrentLang` function has been added (#17222, @nvdaes)
+* Retrieving the "labeledBy" property now works for objects in applications implementing the "labelled-by" IAccessible2 relation and no longer triggers an error. (#17436, @michaelweghorn)
```
```suggestion
* Retrieving the `labeledBy` property now works for objects in applications implementing the `labelled-by` IAccessible2 relation and no longer triggers an error. (#17436, @michaelweghorn)
```

seanbudd committed:
Apply suggestions from code review

michaelweghorn committed:
Add comment why import is needed

Without deferred evaluation (s. PEPs mentioned
in the newly added comment):

    $ ./runlint.bat
    Ensuring NVDA Python virtual environment
    call ruff check --fix
    source\NVDAObjects\IAccessible\__init__.py:1165:30: F821 Undefined name `IAccessible`
         |
    1163 |         return True
    1164 |
    1165 |     def _get_labeledBy(self) -> IAccessible | None:
         |                                 ^^^^^^^^^^^ F821
    1166 |         label = self._getIA2RelationFirstTarget(IAccessibleHandler.RelationType.LABELLED_BY)
    1167 |         if label:
         |

    Found 1 error.
    Deactivating NVDA Python virtual environment

seanbudd committed:
Apply suggestions from code review

seanbudd commented on source/NVDAObjects/IAccessible/__init__.py:
```diff
@@ -3,13 +3,13 @@
 # This file is covered by the GNU General Public License.
 # See the file COPYING for more details.
 
+
```
```suggestion
```

seanbudd committed:
Update source/NVDAObjects/IAccessible/__init__.py

seanbudd committed:
Merge branch 'master' into michaelweghorn/iaccessible2_labelled_by

seanbudd commented on source/NVDAObjects/IAccessible/__init__.py:
```diff
@@ -1162,13 +1161,20 @@ def isPointInObject(self, x, y):
 		return True
 
-	def _get_labeledBy(self):
+	def _get_labeledBy(self) -> "IAccessible" | None:
```
```suggestion
	def _get_labeledBy(self) -> "IAccessible | None":
```

seanbudd committed:
Update source/NVDAObjects/IAccessible/__init__.py

michaelweghorn commented:
Thanks for the review and bringing this in line with the NVDA code conventions, @seanbudd !
